### ISSUE TITLE
Export flag for glyphs?

### ISSUE THREAD

**Issue by madig** [unified-font-object/ufo-spec#68]
I'd like to see a standardized export flag for glyphs, i.e. if a glyph should make it from the source to the binary.

We recently had a situation where we wanted to ship a preview font to a customer while several glyphs were still being worked on. Since the `com.schriftgestaltung.Glyphs.Export` lib key isn't standardized, it made several contortions in our pipeline necessary.

**Comment by anthrotype** [comment]
+1 for a `public.export` (boolean, true by default) key in the GLIF lib

**Comment by madig** [comment]
Then it can be used inside ufo2ft instead of fontmake subsetting the font afterwards.

**Comment by anthrotype** [comment]
actually, I think it's enough to have a single global array of strings, e.g. called `public.subsetGlyphs`, stored in the font's lib.plist, similar to the `public.glyphOrder` list.
I don't think we need to support both a font-level list of "Keep Glyphs" and a glyph-level "Export" flag, like Glyphs.app does. Even there, the two methods are alternative to each other (with the global list taking precedence over the individual glyph export flags).
Usually for a font compiler that has to decide whether to keep or drop some glyphs, it's more important to have the full set of glyphs to keep, than having to loop over each single glyph to get to that.

So my final vote goes for supporting a single `public.subsetGlyphs` in font lib.

**Comment by anthrotype** [comment]
the proposed `public.subsetGlyphs` list may contain glyph names that are not in the `contents.plist`, in which case nothing happens (they don't exist, they can't be kept, simply ignored).

**Comment by chrissimpkins** [comment]
> I don't think we need to support both a font-level list of "Keep Glyphs" and a glyph-level "Export" flag, like Glyphs.app does. Even there, the two methods are alternative to each other (with the global list taking precedence over the individual glyph export flags).
Usually for a font compiler that has to decide whether to keep or drop some glyphs, it's more important to have the full set of glyphs to keep, than having to loop over each single glyph to get to that

+1

I'll add that it would be nice to take these subset list definitions out of XML/plist format and place them in a more human readable and modifiable format that allows for the definitions to be set easily outside of a font editor tool, and perhaps even allows for multiple subsets to be defined within the same file.  A simple newline delimited file would be ideal to support single subsets.  TOML/YAML or something along those lines would be ideal to define more than one subset in a given file and then compilers take a flag with subset name to parse the proper definition from the subset spec file.

**Comment by madig** [comment]
@chrissimpkins Hm, I think your wish is somewhere outside the scope of this proposal. This is more about not exporting glyphs that aren't finished yet or design parts that shouldn't land in the final binaries.

See pyftsubset for something better suited for your use case.

**Comment by chrissimpkins** [comment]
@madig Thanks Nikolaus. Probably better suited as a separate issue report (somewhere, not clear whether compiler or source spec is best spot for this).  This would address your issue for the need to occasionally create different custom subsets from a larger set of glyph designs though, no?  Are you specifying this in the Glyphs editor?

**Comment by madig** [comment]
If you want to create specific subsets for different targets, `pyftsubset` will do what you want -- it's the frontend to fontTools' Subset class. Have a look at the help text. We use txt files of glyph names we want in specific subsets and then repeatedly run pyftsubset on the source binary :) We leave the masters fully featured, the export flag is only denied to glyphs that are still WIP.

**Comment by chrissimpkins** [comment]
@madig ty!

**Comment by madig** [comment]
Any opinions?

**Comment by typesupply** [comment]
> Any opinions?

Would `public.subsetGlyphs` list the glyphs that should be exported? In other words, if a glyph isn't in that list it should be ignored? If that's the case, I don't like that. It should be the other way around: there should be a list of glyphs that should be ignored. Otherwise, every UFO in existence will now generate empty binaries. A single yes/no flag for use in all cases seems pretty blunt. What if something shouldn't be exported in Type 1, but should be in OTF? It's also a bit vague. What is the behavior when a glyph being exported contains a component that references a glyph that  is not being exported? I'm certainly not opposed to the idea, it just there is still a lot to work out here for this to be a part of the spec. 😃

**Comment by typemytype** [comment]
what is wrong with the current `public.glyphOrder`?

following the spec: http://unifiedfontobject.org/versions/ufo3/lib.plist/#publicglyphorder

> The glyph order is stored as a list of glyphs names. Glyph names must not occur more than once. Glyph names in the font may not appear in the order. The order may contain glyphs that are not in the font.

Changing spec should be considered wisely and not randomly cause its needed in a single place.

**Comment by madig** [comment]
> what is wrong with the current public.glyphOrder?

I didn't know `public.glyphOrder` could be used for that?

> What is the behavior when a glyph being exported contains a component that references a glyph that is not being exported?

Glyphs decomposes the non-exported glyph everywhere it's used, pyftsubset keep the non-exported glyph but makes it inaccessible from the cmap iirc. The disadvantage there is that your private glyph-building part components are kept in the final file.

> Otherwise, every UFO in existence will now generate empty binaries. 

Unless the default is implicitly equal to `public.glyphOrder` I guess?

**Comment by typesupply** [comment]
> Glyphs decomposes the non-exported glyph everywhere it's used, pyftsubset keep the non-exported glyph but makes it inaccessible from the cmap iirc. The disadvantage there is that your private glyph-building part components are kept in the final file.

This is where things get thorny and need to be specified very clearly. The spec can't leave things up to interpretation unless it says as much. (See the font.info naming descriptions for an example of deliberate vagueness.)

> Unless the default is implicitly equal to public.glyphOrder I guess?

I think if there is such a list, it should contain what should be excluded. That's backwards compatible.

**But…** The Holy Grail of font formats is to have a single source that can output *all* formats. One of the (minor) goals of UFO in the beginning was just that. It's not defined in the spec because it's a deeeeeep issue. This is the structure that I have had bouncing around in my head for years (rough sketch):

```python
["public.formatExportData.<formatName>"] : {
    "global" : { # settings to be used when exporting formatName unless there are local overrides.
        # format specific key/value pairs
    },
    "local" : {
        "title" : {
            # format specific key/value pairs
        }
    }
}
```

This would allow for multiple formats and multiple output targets per format. (In other words, if I need to subset a single UFO into multiple TTFs, I can have settings for all TTFs and then specific overrides for "my subset 1", "my subset 2", etc. Two problems with this:

1. Specifying the key/value pairs is going to be a lot of work.
2. Is this really something that belongs in the font source? Should it be at a higher level? Is this too workflow specific?

And that's why the spec doesn't get deep into these output details.

**Comment by madig** [comment]
Hm, maybe this can be re-framed. I personally am less concerned with _output details_, e.g. our build system takes care of cutting full WOFFs up into subsets, etc. using plain .txt files as input. I agree that putting build system details into UFOs is probably opening Pandora's box.

I see this hypothetical `public.skipExportGlyphs` list as an _input detail_, as in, how can I stop certain glyphs from reaching the compiler, the file or at least the cmap table?

The primary use-cases where I wished we had something like this have been:
- A designer starts a new project and makes more glyphs than what she needs to send out for review for whatever reason. To avoid awkward "but this other letter here is different from what you sent before!" situations, she would like certain glyphs to just not show up in the file.
- A slight variation on this is when you're using components to build other glyphs but don't want those building blocks to show up in the final file. Here, prescribing decomposing it everywhere it is used sounds like the obvious solution to me. If you don't mind the glyph showing up in the final file, you don't put it into `public.skipExportGlyphs`.
- Two teams work on two scripts in parallel but a customer is only getting one of the scripts as a preview because the other one still needs work. This can be worked around, but is always more tedious than just checking a box in an editor.
- A designer is tasked with taking an existing family with lots of non-compatible outlines and turning it into a variable font with a limited glyph set/script support to start things off. Deleting out-of-scope scripts is possible, but then they need to be merged back in if the scope expands. Testing compatibility and general variation work is much easier if you can quickly mark glyphs you haven't worked on yet as "Do not export" and the compiler doesn't even see them to complain about compatibility.

Regarding the last point, it is important to also consider a Designspace-level `public.skipExportGlyphs` and how e.g. fontmake should treat a UFO-level `public.skipExportGlyphs` when being passed a Designspace. Maybe `public.skipExportGlyphs` can stay in individual UFOs but if trying to interpolate, the compiler must demand that the list matches between all interpolation sources? I.e. when the user clicks "Do not export" on a glyph, the editor must make the distinction between working on a single master UFO and working with master UFOs in a Designspace.

Side note: designspaceLib has a "muteGlyphNames" attribute on sources, but this will just not use the glyph data of a particular source. If you put it on all sources, you'll simply get an empty glyph.

Regarding `public.glyphOrder` again, the specification treats this only as an order and it is handled as such in Trufont and our Fontlab 5 setup (glyphs not present in there are appended to the end in whatever order). I would like to e.g. keep the order of glyphs in the editor but just not export certain glyphs.

**Comment by gferreira** [comment]
two things come to mind:

- what if an ignored glyph is used in an OpenType feature?
- maybe this can be solved more elegantly using layers? the ‘ignore’ flag is a sort of binary mask on top of the foreground glyphs…

**Comment by LettError** [comment]
@madig do you use `public.skipExportGlyphs` in desingspaces?

**Comment by madig** [comment]
> what if an ignored glyph is used in an OpenType feature?

Then the build process could fail with an error pointing out that a glyph is not being exported. Not sure how to handle this more gracefully.

> maybe this can be solved more elegantly using layers? the ‘ignore’ flag is a sort of binary mask on top of the foreground glyphs…

I.e. a hypothetical `public.skipExportGlyphs` layer instead of a lib list? Possible, but then reading the ignore flags would incur disk access to lazy load the layer? And we'd still have deal with the Designspace case and keeping the information in sync between UFOs of a Designspace.

> do you use public.skipExportGlyphs in desingspaces?

No, but I'd like to to ease the pain in the last use-case in my comment above. Why?

**Comment by LettError** [comment]
Something like this:
* `public.skipExportGlyphs` in a designspace.lib: would be used for building a variable font and generating UFO instances without those glyphs.
* A `public.skipExportGlyphs` in a ufo lib: this would tell a font generator to make a binary without those glyphs

--?
In which situation would there be a conflict between a UFO's list and a designspace list?

**Comment by madig** [comment]
I suppose that would work, as in, if you fontmake a Designspace file, the lib key in there is honored and the UFO lib keys are ignored completely; if you fontmake a bunch of UFOs, the lib key in them is honored.

I mainly foresee accidental conflicts as UFOs and Designspaces are conceptually different things that tools need to account for. I.e. you work on a family in one editor and export a layer as a single UFO, an editor or script might mistakenly transfer the list and when you merge it back in, the contents of the list might disagree and lead to support requests. This is hypothetical, but we're specifying the same information in different places with slightly different semantics, which makes me somewhat uneasy.

Also:
1. Which source gets the `public.skipExportGlyphs` key? The one with `info=1`? 
2. Make it possible to use it on instances to filter certain glyphs out? I can't immediately think of a use-case for that though and maybe this is an _output detail_ better taken care of elsewhere...

Maybe it's a good idea to first try to implement this on a Designspace level to get a feel of what needs to be done. Maybe yet another kwarg in all public ufo2ft functions with a list of glyphs to ignore and then just filtering them out of the to-be-compiled layer is enough. @moyogo @anthrotype Let's have a look.

**Comment by LettError** [comment]
>  if you fontmake a Designspace file, the lib key in there is honored and the UFO lib keys are ignored completely

I'm confused. Those UFO lib keys - these would be in the masters?

**Comment by madig** [comment]
Hm? I understood your comment as having two `public.skipExportGlyphs` keys, one on the UFO level for when you are compiling single-master, standalone UFOs and another on the Designspace level when you use UFOs as your masters/sources. Or am I getting confused?

**Comment by LettError** [comment]
No, your summary is correct, that will work.

**Comment by benkiel** [comment]
I do think that this would be handy/useful, and like @LettError's idea of one in the Designspace side and one in the ufo side. For the UFO spec, it would just need to specify what happens at the UFO level, the Designspace spec could specify what to do there when the UFO has a `public.skipExportGlyphs` list. One would need to clearly specify what happens with components, groups, and kerning when glyphs referenced in those are skipped.

> Which source gets the public.skipExportGlyphs key? The one with info=1?

My understanding from the above is that the Designspace file would get the `public.skipExportGlyphs` key, so that you could keep the UFO level and the Designspace level separate. Did I not understand that correctly?

As for the issue of features, that @gferreira pointed out, I don't see that as an issue, as one can already write feature code that references glyphs that aren't in the UFO: the build just fails.

**Comment by madig** [comment]
I have implemented a prototype in https://github.com/googlei18n/ufo2ft/pull/318. The behavior currently is:

1. Components of skipped glyphs are decomposed before the glyphs are dropped from the glyph set.
2. Groups are pruned of all list entries that mention a skipped glyph (might leave behind empty groups)
3. Kerning entries that reference a skipped glyph are dropped
4. ufo2ft's `compile*` functions that take one or more UFO(s) may consult the UFO's `public.skipExportGlyphs` lib key, the `compile*FromDS` functions that take a Designspace may consult only the Designspace's lib key.
5. Feature code is currently untouched.

**Comment by benkiel** [comment]
@madig Empty groups will certainly be a problem, say for sure in kerning (or even then 1 member groups, but that's tolerable). Should it just fail and warn, or should it go deeper to fix the kerning? (though, this is a bit closer to then rewriting features)

**Comment by madig** [comment]
Oh, I didn't think of the effects on kerning. Sure, then let's prune empty groups and remove them from kerning, too. Feature file remains untouched then, most users should have kerning in kerning.plist, no?

**Comment by benkiel** [comment]
Agree that we leave feature file untouched. That is up to the user to keep in sync; that's been the long standing philosophy

**Comment by benkiel** [comment]
And, to be very clear, we should state that the modification does not touch the UFO, it is all on the tool ingesting the UFO.

**Comment by madig** [comment]
The PR at https://github.com/googlei18n/ufo2ft/pull/318 now conforms to the amended https://github.com/unified-font-object/ufo-spec/pull/84. Any more comments? I'd like to go ahead with the implementation 😊

**Comment by madig** [comment]
Merged into the spec and implemented in ufo2ft 🙌
